### ISSUE TITLE
Bug: AES 256 with long passwords does not produce readable documents

### ISSUE THREAD

**Issue by dominicschaub** [qpdf/qpdf#195]
Running 

qpdf --encrypt LONGPASSWORD LONGPASSWORD 256 -- unencrypted.pdf encrypted.pdf

with LONGPASSWORD of e.g. 30 characters produces a PDF that cannot be decrypted by Evince or Okular. Shortening the password to e.g. 15 characters seems to suppress the bug, as does switching to 128 bits security.

**Comment by jberkenbilt** [comment]
Can you see whether creating a file with a long password using some other tool that is capable of encrypting PDF files has this same effect? I actually received a PDF file encrypted with a long password in my day job that was generated by some other tool (not qpdf), which I could not open with evince. I then used qpdf to decrypt the file and opened the resulting file in evince. I also did not try to open it with any other software.

**Comment by dominicschaub** [comment]
Hi, I apologize, I should have used softer language and described my issue as a _possible_ bug. It would appear that the implementation of AES is shockingly buggy across the majority of commonly available software. I evaluated my test pdf (created with QPDF) on a few more readers on various Linux systems and can report the following results:

Adobe Acrobat 9: Crashes the program during opening
MuPDF: Fails with "warning: aes padding out of range"
Evince: Iterates the password prompt ad infinitum (as if the provided password were invalid)
Okular: Again, iterates the password prompt ad infinitum
Foxit Reader: Works perfectly.
Xpdf: Works perfectly.

**Comment by jberkenbilt** [comment]
The spec on AES256 in PDF is written very poorly, and there are parts of it that are ambiguous. When I implemented it in qpdf, I got a trial of Acrobat, generated hundreds of files with random data, wrote code to ensure that all boundary conditions I could think of were exercised, and ran with all sorts of different possible interpretations of the spec. One of the two or three reasonable interpretations worked with all of the test files I generated with Acrobat. None of the others did, though most of them worked with most of the files. I am not at all surprised that many of the other readers crashed or failed. As far as I know, qpdf's AES256 implementation is correct, and qpdf is a useful tool for decrypting and/or reencrypting files that other readers can't open. qpdf is also one of very few PDF readers that can handle an interim AES256 format whose spec was never released by Adobe but was present in a few versions of Adobe. To my knowledge, qpdf is the only package than can also write this deprecated and unsupported format.

Anyway, no worries about the tone. I didn't interpret it in a negative way. I appreciate your taking the time to report this, and if it did turn out to be a qpdf bug, I would want to fix it. However, in this case, I think it's more likely to be a bug in the other packages. If you agree, feel free to close. If you think it's a qpdf bug, we can investigate further.

Thanks for taking the time to help make qpdf better.
